metals/src/main/scala/scala/meta/internal/metals/PackageProvider.scala

```diff
@@ -361,31 +372,23 @@ class PackageProvider(
     refProvider.references(params).flatMap(_.locations)
   }
 
-  private def calcPkgEdit(
-      oldPackageSplit: Vector[Vector[String]],
-      newPackageParts: Vector[String],
-      oldPath: AbsolutePath,
-      packages: Vector[Pkg],
-  ): WorkspaceEdit =
-    packages match {
-      case firstPkg +: otherPkgs =>
-        val lastPkg = otherPkgs.lastOption.getOrElse(firstPkg).ref.pos
-        val range = firstPkg.ref.pos.toLsp.copy(
-          startCharacter = 0,
-          endLine = lastPkg.endLine,
-          endCharacter = lastPkg.endColumn,
-        )
-        val newSplit = splitPackageStatements(oldPackageSplit, newPackageParts)
-        val pkgStatement =
-          newSplit.map(ps => s"package ${ps.mkString(".")}").mkString("\n")
-        if (oldPackageSplit != newSplit)
-          workspaceEdit(oldPath, pkgStatement, range)
-        else new WorkspaceEdit()
-      case _ => new WorkspaceEdit()
+  private def findPkgs(tree: Tree): PkgsStructure = {
+    @tailrec
+    def extractOuterPkg(
+        tree: Tree,
+        acc: PkgsStructure = PkgsStructure.empty,
+    ): PkgsStructure =
+      tree match {
+        case p @ Pkg(_, Nil) => acc.addPkg(p)
+        case p @ Pkg(_, st :: _) => extractOuterPkg(st, acc.addPkg(p))
+        case p: Pkg.Object => acc.withObject(p)
+        case _ => acc
+      }
+    tree match {
+      case Source(List(pk: Pkg)) => extractOuterPkg(pk).reverse
```

Comment:
What about multiple toplevel packages? Does it make sense to support them?

Reply:
Yeah, for split packages, people use those.
```
package scala.meta
package internal.metals
```

Reply:
I was thinking about case like 

```scala
package a {
...
}

package b {
...
}

```

Then `Tree` would be `Source(List(package a, package b))`?

Reply:
Oh, I see. Nah. How would we support this? Which package you'd like to rename? First one? Search for the one that matches the path? Sure, we can make some arbitrary decisions in such a case but I don't think it's worth it. It's not a common case and  I think, that if someone does that it's fair to make fixing packages their responsibility.